plugins/asus-hid/fu-asus-hid.rs

```diff
@@ -0,0 +1,165 @@
+    FlashTaskSomething2 = 0xd000,
+    SwitchToRom = 0xd2, // OEM_MainSwitchToRom, doesn't appear in packet cap
+}
+
+//Info stuff happens on report ID 0x5A
+//Flashing seems to happen on report ID 0x0
+
+#[repr(u8)]
+enum FuAsusHidReportId {
+    Flashing = 0x00,
+    Info = 0x5A,
+}
+
+#[derive(New)]
+struct FuStructAsusManCommand {
+    report_id: FuAsusHidReportId == Info,
+    data: [char; 14] == "ASUS Tech.Inc.",
+    terminator: u8 == 0,
+}
+
+#[derive(New, Getters)]
+struct FuStructAsusManResult {
+    report_id: FuAsusHidReportId == Info,
+    data: [char; 31],
+}
+
+#[derive(New, Getters)]
+struct FuStructAsusHidCommand {
+    report_id: FuAsusHidReportId == Info,
+    cmd: u32,
+    length: u8,
+}
+
+#[derive(New, Getters)]
+struct FuStructAsusHidResult {
+    report_id: FuAsusHidReportId == Info,
+    data: [u8; 31],
+}
+
+#[derive(New, Getters, ParseStream)]
+struct FuStructAsusHidDesc {
+    fga: [char; 8],
+    reserved: u8,
+    product: [char; 6],
+    reserved: u8,
+    version: [char; 8],
+    reserved: u8,
+}
+
+#[derive(New, Parse)]
+struct FuStructAsusHidFwInfo {
+    header: FuStructAsusHidCommand,
+    reserved: u8,
+    description: FuStructAsusHidDesc,
+}
+
+#[derive(New, Getters)]
+struct FuStructAsusPreUpdateCommand {
+    report_id: FuAsusHidReportId == Info,
+    cmd: u32,
+    length: u8,
+    data: [u8; 58],
+}
+
+// Flashing sequence
+
+#[repr(u8)]
+enum FuAsusHidFlashCommand {
+    IdentifyFlash = 0xf0200007,
+    ReadFlash = 0xd1,
+    ClearRemoteBuffer = 0xc0,
+    WritePage = 0xc1,
+    Flash4 = 0xc2,
+    FlushPage = 0xc3,
+    VerifyPage = 0xd0,
+    FlashReset = 0xc4,
+}
+
+#[derive(New, Getters)]
+struct FuStructFlashIdentify {
+    command: u8 == 0x07,
+    offset: u24 = 0xf02000,
+    datasz: u8 == 0x2,
+    data: [u8; 58],
+}
+
+#[derive(New, Getters)]
+struct FuStructFlashIdentifyResponse {
+    command: u32,
+    datasz: u8,
+    part: u16,
+}
+
+#[derive(New, Getters, Setters)]
+struct FuStructAsusReadFlashCommand {
+    command: u8 == 0xd1,
+    offset: u24,
+    datasz: u8,
+    data: [u8; 58],
+}
+
+#[derive(New, Getters, Setters)]
+struct FuStructAsusWriteFlashCommand {
+    command: u8 == 0xc1,
+    offset: u16,
+    datasz: u8,
+    data: [u8; 59],
+}
+
+#[derive(New)]
+struct FuStructAsusFlashReset {
+    command: u8 == 0xc4,
+    reserved: [u8; 62],
+}
+
+#[derive(New)]
+struct FuStructAsusFlushPage {
+    command: u8 == 0xc3,
+    address: u32,
+    page_size: u16 == 0x400,
+}
+
+#[derive(New)]
+struct FuStructAsusClearBuffer {
+    command: u8 == 0xc0,
+    reserved: [u8; 62],
+}
+
+#[derive(New)]
+struct FuStructAsusVerifyBuffer {
+    command: u8 == 0xd0,
+    reserved: [u8; 62],
+}
+
+#[derive(New, Getters)]
+struct FuStructAsusVerifyResult {
+    command: u8 == 0xd0,
+    datasz: u8 == 0x6,
+}
+
+// Wire shark filter
+// !(usbhid.setup.ReportID == 13) && _ws.col.protocol == "USBHID" && !(_ws.col.info == "SET_IDLE Request") && !(_ws.col.info == "SET_IDLE Response")
```

Comment:
maybe stick into the README?

Reply:
I'll drop this when I'm ready to exit draft.  It's just notes for me for now.